File: tools/mesh/uhbd_asc2bin.f
c
c Provided by Dave Sept, modified from UHBD utility
c 
c Converts grid from UHBD ASCII format to UHBD binary format
c
c     iunit        i     unit number to use for writing.
c     iform        i     0=unformatted file, 1=formatted file.
c     title        i     title for file.
c     im           i     numbers of points in x - direction.
c     jm           i     numbers of points in y - direction.
c     km           i     numbers of points in z - direction.
c     h            i     distance between grid points.
c     ox \             / coordinate of grid point (0,0,0)
c     oy  >        i  <  such that the coordinates of (i,j,k) are
c     oz /             \ ( ox+h*i, oy+h*j, oz+h*k).
c     grdflg       i     flag indicating the type of grid.
c     grid         i     grid of values to be printed out.
c     scale        i     factor multiplied times the grid that is output.
c
      integer im, jm, km, grdflg, mxgrd, one, kk
      parameter (mxgrd=129)
      integer iunit, iform
      character*72 title
      real h, ox, oy, oz, grid(mxgrd,mxgrd,mxgrd), scale
      real x0, y0, z0, x1, y1, z1, xmax, ymax, zmax, 
     &     xstart, ystart, zstart, xend, yend, zend
  
c     dum#         dummy real variables
c     idum#        dummy integer variables
c
      real dum2, dum3, dum4, dum5, dum6, dum7, dum8
      integer idum2, idum3, idum4
c
      integer i, j, k, ounit
c
c     flnm          name of trajectory file.
c     newfle        output DCD file name (stemunit.sub.DCD)
c     stem          output DCD file name stem
c     string
c
      character*20 newfle
      character*80 flnm
      character*4 stem
      character*1 string
c
c Statement functions
c begin code
c
      write(6,100) 
100   format(1x,'Convert UHBD ascii grid to binary')
c
      write(6,130) 
130   format(1x,'Enter grid map file name:')
      read(5,135) flnm
135   format(a)
      write(6,140) flnm
140   format(1x,'Grid input file :',a)
      iunit=11
      open(unit=11, file = flnm, status = 'old')
c
      write(6,145) 
145   format(1x,'Enter output file name:')
      read(5,135) newfle
      write(6,148) newfle
148   format(1x,'Grid output file :',a)
c
       ounit=12
         open(12, file = newfle, status='new', form='unformatted')
c            call uhopen(12, newfle, 0)
c
c
        read (iunit,77) title, scale, dum2, grdflg, idum2,
     &                    km, one, km, im, jm, km,
     &                         h ,      ox ,      oy ,      oz ,
     &                    dum3, dum4, dum5, dum6, dum7,
     &                    dum8, idum3, idum4

        write (ounit) title, scale, dum2, grdflg, idum2,
     &                    km, one, km, im, jm, km,
     &                         h ,      ox ,      oy ,      oz ,
     &                    dum3, dum4, dum5, dum6, dum7,
     &                    dum8, idum3, idum4
c
        do  20  k = 1 , km
          read(iunit,78) kk, im, jm
          read(iunit,79) ( ( grid(i,j,k), i=1,im), j=1,jm )
   20   continue
c
      do  30  k = 1 , km
          write(ounit) kk, im, jm
          write(ounit) ( ( grid(i,j,k), i=1,im), j=1,jm )
   30 continue
c
      close(iunit)
      close(ounit)
c
c
c               Format statements.
c
   77 format(a72/2e12.6,5i7/3i7,4e12.6/4e12.6/2e12.6,2i7)
   78 format(3i7)
   79 format( (6(1x,e12.6)) )
      end

File: tools/mesh/uhbd_asc2bin.c
#include <stdio.h>

#include "apbs.h"

#define MXGRD 129

int main(int argc, char **argv) {

    int im, jm, km, grdflg, one, kk;
    double h, ox, oy, oz;
    double scale;
    double dum2, dum3, dum4, dum5, dum6, dum7, dum8;
    int idum2, idum3, idum4;
    int i, j, k;

    char newfile[256];
    char flnm[256];
    char title[72];

    FILE *inFile = NULL;
    FILE *outFile = NULL;

    double grid[MXGRD * MXGRD * MXGRD];
    MAT3(grid, MXGRD, MXGRD, MXGRD);

    printf("Convert UHBD ascii grid to binary\n");
    printf("---------------------------------\n");

    printf("Enter grid map file name: ");
    scanf("%s", flnm);

    inFile = fopen(flnm, "r");
    if( inFile == NULL )
    {
        printf("Couldn't open output file: %s\n", flnm);
        return -1;
    }



    printf("Enter output file name: ");
    scanf("%s", newfile);

    outFile = fopen(newfile, "wb");
    if( outFile == NULL )
    {
        printf("Couldn't open output file: %s\n", newfile);
        return -2;
    }


    /* Read in the DX regular positions */
    /* Get "object" */
    fscanf(inFile, "%s", title);
    fwrite(title, 1, 72, outFile);

    fscanf(inFile, "%lf", &scale);
    fwrite(&scale, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum2);
    fwrite(&dum2, 1, sizeof(double), outFile);

    fscanf(inFile, "%d", &grdflg);
    fwrite(&grdflg, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &idum2);
    fwrite(&idum2, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &km);
    fwrite(&km, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &one);
    fwrite(&one, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &km);
    fwrite(&km, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &im);
    fwrite(&im, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &jm);
    fwrite(&jm, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &km);
    fwrite(&km, 1, sizeof(int), outFile);

    fscanf(inFile, "%lf", &h);
    fwrite(&h, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &ox);
    fwrite(&ox, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &oy);
    fwrite(&oy, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &oz);
    fwrite(&oz, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum3);
    fwrite(&dum3, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum4);
    fwrite(&dum4, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum5);
    fwrite(&dum5, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum6);
    fwrite(&dum6, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum7);
    fwrite(&dum7, 1, sizeof(double), outFile);

    fscanf(inFile, "%lf", &dum8);
    fwrite(&dum8, 1, sizeof(double), outFile);

    fscanf(inFile, "%d", &idum3);
    fwrite(&idum3, 1, sizeof(int), outFile);

    fscanf(inFile, "%d", &idum4);
    fwrite(&idum4, 1, sizeof(int), outFile);

    for (k=0; k<km; k++) {

        fscanf(inFile, "%d", &kk);
        fwrite(&kk, 1, sizeof(int), outFile);

        fscanf(inFile, "%d", &im);
        fwrite(&im, 1, sizeof(int), outFile);

        fscanf(inFile, "%d", &jm);
        fwrite(&jm, 1, sizeof(int), outFile);

        for (j=0; j<jm; j++) {

            for (i=0; i<im; i++ ) {

                fscanf(inFile, "%lf", RAT3(grid, i, j, k));
                fwrite(RAT3(grid, i, j, k), 1, sizeof(double), outFile);

            }
        }
    }

    return 0;
}
